network/crl.go

```diff
@@ -0,0 +1,380 @@
+	c.mutex.Lock()
+	c.cache.Remove(key)
+	c.mutex.Unlock()
+	log.Debugf("CRL: LRU cache: removed '%s'", key)
+	return nil
+}
+
+// DefaultCRLVerifier is a default implementation of CRLVerifier
+type DefaultCRLVerifier struct {
+	Config      CRLConfig
+	Client      CRLClient
+	Cache       CRLCache
+	ParsedCache ParsedCRLCache
+}
+
+// Tries to find cached CRL, fetches using v.Client if not found, checks the signature of CRL using issuerCert
+func (v DefaultCRLVerifier) getCachedOrFetch(uri string, issuerCert *x509.Certificate) (*pkix.CertificateList, error) {
+	// First, try v.ParsedCache
+	crl, err := v.ParsedCache.Get(v.Config.uri)
+	if crl != nil {
+		if err != nil {
+			// non-empty result + error, should never happen
+			return nil, err
+		}
+
+		return crl, nil
+	}
+
+	// If not found, search in v.Cache, if found then parse, verify and store in v.ParsedCache
+	rawCRL, err := v.Cache.Get(v.Config.uri)
+	if rawCRL != nil {
+		if err != nil {
+			// non-empty result + error, should never happen
+			return nil, err
+		}
+
+		crl, err := x509.ParseCRL(rawCRL)
+		if err != nil {
+			// Should never happen since we only store parse-able ones in v.Cache
+			log.WithError(err).Warnf("CRL: cache contains invalid CRL")
+			v.Cache.Remove(uri)
+			return nil, err
+		}
+
+		err = issuerCert.CheckCRLSignature(crl)
+		if err != nil {
+			// Should never happen since we only store verified ones in v.Cache;
+			// however, this may happen if wrong CA cert was passed, like first time we check with right
+			// one and store in v.ParsedCache and this time we use a different CA cert and signature check fails
+			log.WithError(err).Warnf("CRL: cache contains CRL with invalid signature")
+			v.Cache.Remove(uri)
+			return nil, err
+		}
+
+		v.ParsedCache.Put(uri, crl)
+
+		return crl, nil
+	}
+
+	// Not found in both caches, gotta fetch
+	rawCRL, err = v.Client.Fetch(uri)
+	if rawCRL != nil {
+		if err != nil {
+			// non-empty result + error, should never happen
+			return nil, err
+		}
+
+		crl, err := x509.ParseCRL(rawCRL)
+		if err != nil {
+			log.WithError(err).Debugf("CRL: Cannot parse CRL from '%s'", uri)
+			return nil, err
+		}
+
+		err = issuerCert.CheckCRLSignature(crl)
+		if err != nil {
+			log.WithError(err).Warnf("CRL: Failed to check signature for CRL at %s", uri)
+			return nil, err
+		}
+
+		v.Cache.Put(uri, rawCRL)
+		v.ParsedCache.Put(uri, crl)
+
+		return crl, nil
+	}
+
+	return nil, err
+}
+
+// Verify ensures configured CRLs do not contain certificate from passed chain
+func (v DefaultCRLVerifier) Verify(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	for _, chain := range verifiedChains {
+		log.Debugf("CRL: Verifying '%s'", chain[0].Subject.String())
+
+		if len(v.Config.uri) == 0 && v.Config.fromCert == crlFromCertIgnore {
+			log.Debugln("CRL: Skipping check since no config URI specified and we were told to ignore URIs from certificate")
+			return nil
+		}
+
+		cert := chain[0]
```

Comment:
what if there empty chain? can we test it in unittests?

Reply:
Well, I guess it's not possible to have completely empty chain in TLS. There should be at least one certificate. Cannot prove it right now, gotta dig into docs. But you gave me an idea: the issuer (`chain[-1]`) may be empty in case of single-certificate chain.